hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,346 @@
+    AnalysisResults {
+        ctx,
+        out_wire_values,
+        in_wire_value: all_results.in_wire_value,
+        case_reachable: all_results.case_reachable,
+        bb_reachable: all_results.bb_reachable,
+    }
+}
+
+fn propagate_leaf_op<V: AbstractValue>(
+    ctx: &impl DFContext<V>,
+    n: Node,
+    ins: &[PV<V>],
+    num_outs: usize,
+) -> Option<ValueRow<V>> {
+    match ctx.get_optype(n) {
+        // Handle basics here. We could instead leave these to DFContext,
+        // but at least we'd want these impls to be easily reusable.
+        op if op.cast::<MakeTuple>().is_some() => Some(ValueRow::from_iter([PV::new_variant(
+            0,
+            ins.iter().cloned(),
+        )])),
+        op if op.cast::<UnpackTuple>().is_some() => {
+            let elem_tys = op.cast::<UnpackTuple>().unwrap().0;
+            let tup = ins.iter().exactly_one().unwrap();
+            tup.variant_values(0, elem_tys.len())
+                .map(ValueRow::from_iter)
+        }
+        OpType::Tag(t) => Some(ValueRow::from_iter([PV::new_variant(
+            t.tag,
+            ins.iter().cloned(),
+        )])),
+        OpType::Input(_) | OpType::Output(_) | OpType::ExitBlock(_) => None, // handled by parent
+        OpType::Call(_) => None,  // handled via Input/Output of FuncDefn
+        OpType::Const(_) => None, // handled by LoadConstant:
+        OpType::LoadConstant(load_op) => {
+            assert!(ins.is_empty()); // static edge, so need to find constant
+            let const_node = ctx
+                .single_linked_output(n, load_op.constant_port())
+                .unwrap()
+                .0;
+            let const_val = ctx.get_optype(const_node).as_const().unwrap().value();
+            Some(ValueRow::single_known(
+                1,
+                0,
+                partial_from_const(ctx, n, const_val),
+            ))
+        }
+        OpType::LoadFunction(load_op) => {
+            assert!(ins.is_empty()); // static edge
+            let func_node = ctx
+                .single_linked_output(n, load_op.function_port())
+                .unwrap()
+                .0;
+            // Node could be a FuncDefn or a FuncDecl, so do not pass the node itself
+            Some(ValueRow::single_known(
+                1,
+                0,
+                ctx.value_from_function(func_node, &load_op.type_args)
+                    .map_or(PV::Top, PV::Value),
+            ))
+        }
+        OpType::ExtensionOp(e) => {
+            // Interpret op using DFContext
+            let init = if ins.iter().contains(&PartialValue::Bottom) {
+                // So far we think one or more inputs can't happen.
+                // So, don't pollute outputs with Top, and wait for better knowledge of inputs.
+                PartialValue::Bottom
+            } else {
+                // If we can't figure out anything about the outputs, assume nothing (they still happen!)
+                PartialValue::Top
+            };
```

Comment:
Yeah, coverage comes in #1603 .

`interpret_leaf_op` can do anything, it gets an `&mut [PartialValue<V>]` so can overwrite, it does not have to use join.

Note that returning `Bottom` erroneously is unsafe, i.e. can lead the analysis to conclude things that are not true. Returning `Top` is conservative.

Reply:
The analysis could be better than `contains(PartialValue::Bottom)`, i.e. we could apply the `Bottom` case more often, but at least this way is conservative.

(e.g. `Sum(1, [Bottom, foo, bar])` also "can't happen", so must result from not-yet-computed data that'll turn up later in the analysis - that is, for a PartialSum, we can discount any variant at least one of whose value's (recursively) "contains bottom", and then the PartialSum classed as containing-bottom if it has no variants left)

Reply:
Oooh, but, as you suggest [here](https://github.com/CQCL/hugr/pull/1664/files#r1846407383), maybe we can simplify.

So there are some leaf ops that can figure out properties of their outputs even when *some* of the inputs are completely unknown, and I had thought that that meant we couldn't do as you suggest, but actually....maybe we can; we'll call `interpret_leaf_op` again later when we have non-bottom inputs. In which case we end up
1. `interpret_leaf_op` never sees any `Bottom` inputs
2. initial outputs passed to `interpret_leaf_op` are always `PartialValue::Top`
3. if any input is `Bottom`, all outputs are set to `Bottom` (without even calling `interpret_leaf_op`)

Does that actually work?!?
This is partly about being clear about when Bottom is used - i.e.  only as part of initialization, as a means of bootstrapping cyclic dependencies (maybe non-cyclic depending on how "clever" ascent tries to be)

Reply:
Ok, `interpret_leaf_op` no longer called if any input is Bottom. (I considered a default impl that checks row-contains-bottom but I think that if (a) Bottom can only occur during bootstrap/initialisation, and (b) the op doesn't execute unless it gets a valid input row, this is ok. The case of Conditionals being simplifiable because we know their predicate even if not all their inputs, never hits interpret_leaf_op, so I think this is OK.)

I also considered changing it to return `Vec<PartialValue<V>>` rather than take a mutable reference, but this does mean the callee becomes responsible for allocating a vector of the correct length; there are potentially many callees in the future, and only one caller, which can do that work. If you feel strongly that returning `Vec<PartialValue<V>>` would be better, then I'm willing to change.